Some OSSA related optimization improvements

eeckstein opened the pull request:
Several (relatively) small optimization improvements to better deal with OSSA:

* GlobalPropertyOpt: handle `load_borrow` and `destroy_addr`
* COWArrayOpt: handle `load_borrow`
* LoopRotate: handle `copy_value` and `begin_borrow` correctly
* LoopRotate: don't rotate a loop if the new header is loop exiting as well
* LICM: (limited) support for OSSA
* SimplifyDestructure: canonicalize `destructure_tuple` and `destructure_struct` for trivial tuples/structs
* simplify `load_borrow`
* remove `begin_borrow` if the borrowed value is a `thin_to_thick_function`
* run `begin_borrow` simplification in SILCombine

For details see the commit messages.

This is part of fixing benchmark regressions when enabling OSSA modules.
rdar://139773406

eeckstein committed:
GlobalPropertyOpt: handle load_borrow and destroy_addr

eeckstein committed:
COWArrayOpt: handle load_borrow

eeckstein committed:
LoopRotate: handle copy_value and begin_borrow correctly

eeckstein committed:
LoopRotate: don't rotate a loop if the new header is loop exiting as well

This doesn't give any performance benefit.

eeckstein committed:
LICM: (limited) support for OSSA

The first step: allow hoisting instructions which only have trivial operands and results.

eeckstein committed:
SimplifyDestructure: canonicalize destructure_tuple and destructure_struct for trivial tuples/structs

Replace destructure_tuple with tuple_extract instructions and destructure_struct with struct_extract instructions.
This canonicalization helps other optimizations to e.g. CSE tuple_extract/struct_extract.

eeckstein committed:
Optimizer: simplify load_borrow

* Remove dead `load_borrow` instructions (replaces the old peephole optimization in SILCombine)
* If the `load_borrow` is followed by a `copy_value`, combine both into a `load [copy]`

eeckstein commented:
@swift-ci benchmark

meg-gupta commented on SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyBeginAndLoadBorrow.swift:
```diff
@@ -21,7 +21,52 @@ extension BeginBorrowInst : OnoneSimplifyable {
       tryReplaceBorrowWithOwnedOperand(beginBorrow: self, context)
+
+    } else if let thin2thickFn = borrowedValue as? ThinToThickFunctionInst,
+              uses.filterUsers(ofType: BranchInst.self).isEmpty
```
Looks like `ThinToThickFunctionInst` always produces values of none ownership. I am not sure if the borrowedValue can ever be `ThinToThickFunctionInst `. Can you add a test and move this to a separate commit if this is intended ?

meg-gupta commented:
Awesome! I left a minor comment but this LGTM!

meg-gupta commented on lib/SILOptimizer/LoopTransforms/LICM.cpp:
```diff
@@ -887,6 +902,10 @@ void LoopTreeOptimization::analyzeCurrentLoop(
   for (auto *BB : Loop->getBlocks()) {
     SmallVector<SILInstruction *, 8> sideEffectsInBlock;
     for (auto &Inst : *BB) {
+      if (hasOwnershipOperandsOrResults(&Inst)) {
```
Can you ensure there are tests covering other ownership instructions without side-effects ? load_borrow/destructure_struct/destructure_tuple etc

eeckstein replied:
I added a few tests - for instructions which really don't work if hoisted out of the loop.
This check is currently overly conservative. E.g. it is possible to hoist guaranteed values out of the loop.
Let's improve that when we convert the whole LICM optimization to OSSA.

eeckstein committed:
Optimizer: temporarily disable AccessPathVerification in the late pipeline.

It triggers a false alarm when building SwiftDocC on linux
rdar://141270464

eeckstein committed:
Simplification: remove `begin_borrow` if the borrowed value is a `thin_to_thick_function`

`thin_to_thick_function` has "none" ownership and is compatible with guaranteed values.
Therefore the `begin_borrow` is not needed.

eeckstein committed:
Simplification: run begin_borrow simplification in SILCombine

eeckstein commented:
@swift-ci test
